### PR TITLE
Add data-state to modal

### PR DESCRIPTION
**Pull request by philearley**
# Description

Add data state to modal to aid in test running.
This change ensures the data state is updated once all transitions have fininshed.
Selenium was attempting to select dialogs before they had finished rendering on the page, causing false failures.

### CODE REVIEW

File: src/components/modal/modal.js [hunk 1]
```diff
@@ -155,10 +156,14 @@ class Modal extends React.Component {
       // it was added as the Portal library we use messes up the callstack
       // when assigning the ref and triggering componentDidUpdate
       setTimeout(() => {
+        clearTimeout(this.openTimeout);
+        this.openTimeout = setTimeout(() => { this.setState({ state: 'open' }); }, 500);
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: would it be safer to set the state depending on `this.props.open`.

eg. 

```
this.setState({ state: this.props.open ? 'open' : 'closed' })
```

File: src/components/modal/modal.js [hunk 2]
```diff
@@ -112,6 +112,7 @@ class Modal extends React.Component {
      * @type {Boolean}
      */
     this.listening = false;
+    this.state = { state: 'default' };
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: should this be `open` or `closed`? not sure what benefit you get from `default`?

File: src/components/modal/modal.js (hunk 1 above)
Comment (andrewjtait) [reply src/components/modal/modal.js]: (if you do this you can pull it into a reusable function)

File: src/components/modal/modal.js [hunk 3]
```diff
@@ -155,10 +156,14 @@ class Modal extends React.Component {
...
       // when assigning the ref and triggering componentDidUpdate
       setTimeout(() => {
+        clearTimeout(this.openTimeout);
+        this.openTimeout = setTimeout(() => { this.setState({ state: 'open' }); }, 500);
         this.onOpening; // eslint-disable-line no-unused-expressions
       });
       _window.addEventListener('keyup', this.closeModal);
     } else if (!this.props.open && this.listening) {
+      clearTimeout(this.openTimeout);
+      this.openTimeout = setTimeout(() => { this.setState({ state: 'closed' }); }, 500);
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: could `500` be pulled out into a constant in the file? and shared between this and the animations?

File: src/components/modal/modal.js (hunk 2 above)
Comment (philearley) [reply src/components/modal/modal.js]: We have used idle and default in other locations, it felt the correct thing to do, to initialise with default. Happy to set this to default as a closed if it makes more sense.

File: src/components/modal/modal.js [hunk 4]
```diff
@@ -112,6 +113,12 @@ class Modal extends React.Component {
      * @type {Boolean}
      */
     this.listening = false;
+    this.state = { state: 'closed' };
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: iit could be open by default. I'd also add some comments to this to explain what it is used for

File: src/components/modal/modal.js [hunk 5]
```diff
@@ -112,6 +113,12 @@ class Modal extends React.Component {
...
+  }
+
+  updateState = () => {
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: maybe add a description to explain what this functions purpose is?

File: src/components/modal/modal.js [hunk 6]
```diff
@@ -112,6 +113,12 @@ class Modal extends React.Component {
...
+    clearTimeout(this.openTimeout);
+    this.openTimeout = setTimeout(() => { this.setState({ state: this.props.open ? 'open' : 'closed' }); }, TIMEOUT);
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: maybe break this onto a few lines so its easier to read
```js
this.openTimeout = setTimeout(() => {
  this.setState({ state: this.props.open ? 'open' : 'closed' });
}, TIMEOUT);
```

File: src/components/modal/modal.js [hunk 7]
```diff
@@ -155,11 +162,13 @@ class Modal extends React.Component {
       // it was added as the Portal library we use messes up the callstack
       // when assigning the ref and triggering componentDidUpdate
       setTimeout(() => {
+        this.updateState();
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: `updateState` is a weird name inside of a React component - maybe it should be something more clear?

File: src/components/modal/modal.js (hunk 4 above)
Comment (philearley) [reply src/components/modal/modal.js]: open is a defaultProp and is set to false.
What would you suggest to be the best wait to mirror the component that loads this?

File: src/components/modal/modal.js [hunk 8]
```diff
@@ -148,18 +172,20 @@ class Modal extends React.Component {
    */
   componentDidUpdate() {
     const _window = Browser.getWindow();
+    // this.setState((prevState, props) => {
+    //   return { counter: prevState.counter + props.step };
+    // });
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: comment

File: src/components/modal/modal.js [hunk 9]
```diff
@@ -148,18 +172,20 @@ class Modal extends React.Component {
...
 
     if (this.props.open && !this.listening) {
       this.listening = true;
-      // TODO: try to remove this with React 16 upgrade with native Portals
-      // it was added as the Portal library we use messes up the callstack
-      // when assigning the ref and triggering componentDidUpdate
```
Comment (andrewjtait) [inline src/components/modal/modal.js]: this is still a valid `TODO`

**Comment by philearley** [comment]
Problematic spec continues to pass after latest changes.

**Pull request merged**